Source/Core/DiscIO/NANDImporter.cpp

```diff
@@ -0,0 +1,160 @@
+    PanicAlertT("NAND Importer: Unexpected filesize. 0x%x does not equal 0x%x.", file.GetSize(),
+                NAND_BIN_SIZE);
+    return false;
+  }
+  m_nand = new u8[NAND_SIZE];
+  m_nand_keys = new u8[NAND_KEYS_SIZE];
+  for (int i = 0; i < NAND_TOTAL_BLOCKS; i++)
+  {
+    file.ReadBytes(m_nand + i * NAND_BLOCK_SIZE, NAND_BLOCK_SIZE);
+    file.Seek(NAND_ECC_BLOCK_SIZE, SEEK_CUR);  // We don't care about the ECC blocks
+  }
+  file.ReadBytes(m_nand_keys, NAND_KEYS_SIZE);
+  file.Close();
+  return true;
+}
+
+void NANDImporter::FindSuper()
+{
+  u32 super = 0;
+  u32 newestVersion = 0;
+  for (int pos = 0x1fc00000; pos < NAND_SIZE; pos += 0x40000)
+  {
+    if (Common::swap32(m_nand + pos) == 'SFFS')
+    {
+      u32 version = Common::swap32(m_nand + pos + 4);
+      if (super == 0 || version > newestVersion)
+      {
+        super = pos;
+        newestVersion = version;
+      }
+    }
+  }
+  m_nand_fat = m_nand + super + 0xC;
+  m_nand_fst = m_nand_fat + 0x10000;
+}
+
+std::string NANDImporter::GetPath(NANDFSTEntry* entry, std::string parent_path)
+{
+  std::string name(reinterpret_cast<char*>(entry->name), sizeof NANDFSTEntry::name);
+  // Get rid of any extra null characters
+  while (name.back() == '\0')
+    name.pop_back();
+
+  if (name.front() == '/' || parent_path.back() == '/')
+    return parent_path + name;
+  else
+    return parent_path + "/" + name;
+}
+
+void NANDImporter::ProcessEntry(u16 entry_number, std::string parent_path)
+{
+  NANDFSTEntry* entry = reinterpret_cast<NANDFSTEntry*>(
+      m_nand_fst + sizeof NANDFSTEntry * Common::swap16(entry_number));
+  if (entry->sib != 0xffff)
+    ProcessEntry(entry->sib, parent_path);
+
+  if ((entry->mode & 3) == 1)
+    ProcessFile(entry, parent_path);
+  else if ((entry->mode & 3) == 2)
+    ProcessDirectory(entry, parent_path);
+}
+
+void NANDImporter::ProcessDirectory(NANDFSTEntry* entry, std::string parent_path)
+{
+  std::string path = GetPath(entry, parent_path);
+  File::CreateDir(path);
+
+  if (entry->sub != 0xFFFF)
+    ProcessEntry(entry->sub, path);
+}
+
+void NANDImporter::ProcessFile(NANDFSTEntry* entry, std::string parent_path)
+{
+  std::string path = GetPath(entry, parent_path);
+  File::IOFile file(path, "wb");
+  std::vector<u8> key(16);
+  memcpy(key.data(), m_nand_keys + 0x158, key.size());
+  u32 remaining_bytes = Common::swap32(entry->size);
+  u16 sub = Common::swap16(entry->sub);
+
+  while (remaining_bytes)
+  {
+    std::vector<u8> iv(16, 0);
+    std::vector<u8> block =
+        Common::AES::Decrypt(key.data(), iv.data(), m_nand + 0x4000 * sub, 0x4000);
+    u32 size = remaining_bytes < 0x4000 ? remaining_bytes : 0x4000;
+    file.WriteBytes(block.data(), size);
+    remaining_bytes -= size;
+    sub = Common::swap16(m_nand_fat + 2 * sub);
+  }
+  file.Close();
+}
+
+void NANDImporter::ExtractCertificates(std::string nand_root)
+{
+  File::IOFile file(nand_root + "/title/00000001/0000000d/content/00000011.app", "rb");
+  struct Certificate
```

Comment:
The struct and the certificates[] declaration should probably be separated.

Also, certificates should be a std::array or something.